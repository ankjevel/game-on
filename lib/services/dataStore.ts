import uuid from 'uuid'
import redis from '../adapters/redis'
import parse from '../utils/parse'

import {
  Action,
  CreateGroupInput,
  CreateUserInput,
  GetResult,
  Group,
  isAction,
  isGroup,
  isUser,
  StoreTypes,
  User,
} from '../types/dataStore'
export { StoreTypes, User, Group, CreateGroupInput, CreateUserInput, GetResult }

const client = redis()

const newId = (type: StoreTypes) => `${type}:${uuid.v4()}`

export const create = async <T extends Group | Action | User>(
  type: StoreTypes,
  format: (id: string) => T
): Promise<T> => {
  const id = newId(type)

  if (await client.get(id)) {
    return await create(type, format)
  }

  const object = format(id)

  await client.set(
    id,
    typeof object === 'string' ? object : JSON.stringify(object)
  )

  return object
}

export const update = async <T extends Group | Action | User>(
  id: T['id'],
  data: T
) => {
  if (id !== data.id) {
    return
  }

  const prev = await get({ id })

  if (prev == null) {
    return
  }

  await client.set(id, {
    ...prev,
    ...data,
  })
}

export const get = async ({
  type,
  id,
}: {
  type?: StoreTypes
  id: string
}): Promise<GetResult> => {
  const query =
    id.includes(':') === false
      ? type == null
        ? null
        : `${type}:${id}`
      : type != null && id.split(':')[0] != type
      ? null
      : id

  if (query == null) {
    return null
  }

  const result = await client.get(query)

  if (result === null) {
    return null
  }

  try {
    if (type === StoreTypes.User || id.includes(StoreTypes.User)) {
      const parsed = parse<User>(result)
      return isUser(parsed) ? parsed : null
    }

    if (type === StoreTypes.Group || id.includes(StoreTypes.Group)) {
      const parsed = parse<Group>(result)
      return isGroup(parsed) ? parsed : null
    }

    if (type === StoreTypes.Action || id.includes(StoreTypes.Action)) {
      const parsed = parse<Action>(result)
      return isAction(parsed) ? parsed : null
    }
  } catch (error) {
    console.error(result, error)
    return null
  }

  return null
}
