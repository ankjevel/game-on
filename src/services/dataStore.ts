type Check = <T>(result: MaybeNull<T>) => boolean

import uuid from 'uuid'
import * as userService from './user'
import { parse } from '../utils'
import redis from '../adapters/redis'

import {
  Action,
  GetResult,
  Group,
  isAction,
  isGroup,
  isUser,
  StoreTypes,
  Types,
  User,
} from '../types/dataStore'

export { Action, GetResult, Group, StoreTypes, Types, User, Check }

const client = redis()

export const newId = (type: StoreTypes) => `${type}:${uuid.v4()}`

export const stripTag = (input: string) => input.split(':')[1] || ''

export const checkDuplicate = async <T extends Types>(input: T) => {
  if (isAction(input) || isGroup(input)) {
    return
  }

  if (isUser(input)) {
    return userService.checkDuplicate(input)
  }
}

export const create = async <T extends Types>(
  type: StoreTypes,
  format: (id: string) => T
): Promise<T> => {
  const id = newId(type)

  if (await client.get(id)) {
    return await create(type, format)
  }

  const object = format(id)

  await checkDuplicate<T>(object)

  await client.set(
    id,
    typeof object === 'string' ? object : JSON.stringify(object)
  )

  return object
}

export const all = async (prefix: StoreTypes) =>
  await client.keys(`${prefix}:*`)

export const update = async <T extends Types>(
  id: T['id'],
  data: T,
  type: StoreTypes
) => {
  if (id !== data.id) {
    return
  }

  const prev = await getWrapper({ id, type })
  if (prev == null) {
    return
  }

  await client.set(
    id,
    JSON.stringify({
      ...prev,
      ...data,
    })
  )
}

export const get = async <T extends Types>({
  type,
  id,
  check,
}: {
  type: StoreTypes
  id: string
  check: Check
}): Promise<MaybeNull<T>> => {
  const query =
    id.includes(':') === false
      ? type == null
        ? null
        : `${type}:${id}`
      : type != null && id.split(':')[0] != type
      ? null
      : id

  if (query == null || !query.startsWith(`${type}:`)) {
    return null
  }

  const result = await client.get(query)
  if (result == null) {
    return null
  }

  try {
    const parsed = parse<T>(result)
    return check(parsed) ? parsed : null
  } catch (error) {
    console.error(result, error)
    return null
  }
}

export const getWrapper = async <T extends Types>({
  type,
  id,
}: {
  type: StoreTypes
  id: string
}): Promise<MaybeNull<T>> => {
  let check: Check

  switch (type) {
    case StoreTypes.Action:
      check = isAction
      break
    case StoreTypes.Group:
      check = isGroup
      break
    case StoreTypes.User:
      check = isUser
      break
    default:
      throw new Error('missing type')
  }

  return get<T>({ id, type, check })
}
