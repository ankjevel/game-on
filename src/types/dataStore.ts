import { hasProp, looksLikeEmail, isNumber, nullOrEmpty } from '../utils'
import {
  Action,
  IActionRunning,
  Group,
  User,
  UserWithOutPassword,
  GetResult,
  KeyValue,
} from 'dataStore'

export type Types = Action | Group | User

export enum StoreTypes {
  Action = 'action',
  Group = 'group',
  User = 'user',
  ActionRunning = 'action:running',
}

export type ActionRunning = IActionRunning<NewAction, UserSummary>

export enum NewActionEnum {
  None = 'none',
  Bet = 'bet',
  Check = 'check',
  Call = 'call',
  Raise = 'raise',
  AllIn = 'allIn',
  Fold = 'fold',

  Draw = 'draw',
  Winner = 'winner',
  Back = 'back',
  Bank = 'bank',
  Join = 'join',
  Leave = 'leave',
  SittingOut = 'sittingOut',
}

export type NewAction = {
  type: NewActionEnum
  value?: number
  winners?: User['id'][]
}

export type UserSummary = {
  bet: number
  status: NewActionEnum
}

export const isUserSummary = (any: any): any is UserSummary =>
  any != null &&
  hasProp<any>(any, 'bet') &&
  isNumber(any.bet) &&
  hasProp<any>(any, 'status') &&
  isNewActionType((any as any).status)

export const isNewActionType = (any: any): any is NewActionEnum =>
  Object.values(NewActionEnum).includes(any)

export const isNewAction = (
  any: NewAction | unknown,
  strict = false
): any is NewAction => {
  if (
    !hasProp<NewAction>(any, 'type') ||
    !isNewActionType((any as any).type) ||
    (hasProp<NewAction>(any, 'value') && any.value != null
      ? !isNumber((any as any).value)
      : false)
  ) {
    return false
  }

  if (!strict) {
    return true
  }

  switch (any.type) {
    case NewActionEnum.AllIn:
    case NewActionEnum.Back:
    case NewActionEnum.Bet:
    case NewActionEnum.Call:
    case NewActionEnum.Check:
    case NewActionEnum.Fold:
    case NewActionEnum.None:
    case NewActionEnum.SittingOut:
      return isNumber(any.value) === false
    case NewActionEnum.Draw: // draw COULD include users that went all in.
    case NewActionEnum.Winner:
      return Array.isArray(any.winners)
        ? any.winners.every(winner => checkId(winner, StoreTypes.User))
        : true
    default:
      return isNumber(any.value)
  }
}

export { Action, Group, User, GetResult }

export const checkId = (input: string, type: StoreTypes) =>
  input.split(':')[0] === type.split(':')[0]

export const isUser = (any: User | unknown): any is User =>
  any != null &&
  (hasProp<User>(any, 'id') &&
    !nullOrEmpty(any.id) &&
    checkId(any.id, StoreTypes.User)) &&
  (hasProp(any, 'email') &&
    !nullOrEmpty(any.email) &&
    looksLikeEmail(any.email)) &&
  (hasProp(any, 'password') && !nullOrEmpty(any.password))

export const isUserWithOutPassword = (
  any: UserWithOutPassword | unknown
): any is UserWithOutPassword =>
  any != null &&
  (hasProp<UserWithOutPassword>(any, 'id') &&
    !nullOrEmpty(any.id) &&
    checkId(any.id, StoreTypes.User)) &&
  (hasProp(any, 'email') &&
    !nullOrEmpty(any.email) &&
    looksLikeEmail(any.email))

export const isGroup = (any: Group | unknown): any is Group =>
  any != null &&
  (hasProp<Group>(any, 'id') &&
    typeof any.id === 'string' &&
    !nullOrEmpty(any.id) &&
    checkId(any.id, StoreTypes.Group)) &&
  (hasProp(any, 'name') &&
    typeof any.name === 'string' &&
    !nullOrEmpty(any.name)) &&
  (hasProp(any, 'owner') &&
    typeof any.owner === 'string' &&
    !nullOrEmpty(any.owner)) &&
  (hasProp(any, 'startSum') && isNumber(any.startSum)) &&
  (hasProp(any, 'users')
    ? Array.isArray(any.users) &&
      (any.users as Group['users']).every(
        ({ id, sum }) => checkId(id, StoreTypes.User) && isNumber(sum)
      )
    : true) &&
  (hasProp(any, 'action')
    ? typeof any.action === 'string' && checkId(any.action, StoreTypes.Action)
    : true)

export const isAction = (any: Action | unknown): any is Action =>
  any != null &&
  hasProp<Action>(any, 'id') &&
  typeof any.id === 'string' &&
  checkId(any.id, StoreTypes.Action)

export const isTurn = (
  any: KeyValue<NewActionEnum> | unknown
): any is KeyValue<NewActionEnum> =>
  any != null &&
  Object.entries(any as any).every(
    ([key, value]) => checkId(key, StoreTypes.User) && isNewAction(value)
  )

export const isActionRunning = (
  any: ActionRunning | unknown
): any is ActionRunning =>
  isAction(any) &&
  (hasProp<ActionRunning>(any, 'round') && isNumber(any.round)) &&
  (hasProp(any, 'groupID') && checkId(any.groupID, StoreTypes.Group)) &&
  (hasProp(any, 'queued') && isTurn(any.queued)) &&
  (hasProp(any, 'turn') &&
    Object.entries(any.turn).every(
      ([key, value]) => checkId(key, StoreTypes.User) && isUserSummary(value)
    )) &&
  (hasProp(any, 'button') && checkId(any.button, StoreTypes.User)) &&
  (hasProp(any, 'big') && checkId(any.big, StoreTypes.User)) &&
  (hasProp(any, 'pot') && isNumber(any.round)) &&
  (Array.isArray(any.sittingOut)
    ? any.sittingOut.every(user => checkId(user, StoreTypes.User))
    : true)
