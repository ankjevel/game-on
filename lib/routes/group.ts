import { Route } from 'Route'
import hasProp from '../utils/hasProp'
import isNumber from '../utils/isNumber'
import * as groupService from '../services/group'

export const register: Route = (app, auth) => {
  app.get('/group', async ({ query }, res) => {
    const name: string | undefined =
      hasProp(query, 'name') && typeof query.name === 'string'
        ? query.name.substr(0, 255)
        : undefined

    const startSum: number | undefined =
      hasProp(query, 'startSum') &&
      isNumber(query.startSum) &&
      isNumber(parseInt(query.startSum))
        ? Math.max(
            0,
            Math.min(parseInt(query.startSum), Number.MAX_SAFE_INTEGER)
          )
        : undefined

    res.send(await groupService.newGroup({ name, startSum }))
  })

  app.get(
    '/group/:id/join',
    auth,
    async ({ params: { id }, query, user }, res) => {
      console.log({ id, query, user })

      return res.sendStatus(200)
    }
  )
}
