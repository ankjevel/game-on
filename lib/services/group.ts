import { CreateInput, Group } from 'dataStore'
import { create, StoreTypes } from './dataStore'

export const newGroup = async ({
  name = '',
  startSum = 1000,
}: CreateInput): Promise<Group> => {
  const group = await create<Group>(StoreTypes.Group, id => ({
    id,
    name,
    startSum,
    users: []
  }))

  return group
}
