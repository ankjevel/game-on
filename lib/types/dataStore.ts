import hasProp from '../utils/hasProp'
import { Action, Group, User, CreateInput, GetResult } from 'dataStore'
export { Action, Group, User, CreateInput, GetResult }

export enum StoreTypes {
  Action = 'action',
  Group = 'group',
  User = 'user',
}

export const isUser = (any: any): any is User => {
  return any != null && hasProp(any, 'id')
}

export const isGroup = (any: any): any is Group => {
  return (
    any != null &&
    hasProp(any, 'id') &&
    hasProp(any, 'users') &&
    typeof any.id === 'string' &&
    Array.isArray(any.users) &&
    any.users.every(value => isUser(value)) &&
    ((hasProp(any, 'action') && isAction(any.action)) || true)
  )
}

export const isAction = (any: any): any is Action => {
  return any != null && hasProp(any, 'id') && typeof any.id === 'string'
}
