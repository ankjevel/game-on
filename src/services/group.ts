import {
  ActionRunning,
  Group,
  Order,
  User,
  UserWithOutPassword,
} from 'dataStore'

import * as dataStore from './dataStore'
import { newDeck } from './cards'
import { clone } from '../utils'

const getWrapper = async (id: Group['id']) => {
  return await dataStore.get<Group>({
    id,
    type: 'group',
    check: dataStore.isGroup,
  })
}

const checkIfAlreadyInAGroup = async (id: UserWithOutPassword['id']) => {
  for (const key of await dataStore.all('group')) {
    const res = await getWrapper(key)

    if (res != null && res.users.some(user => user.id === id)) {
      return true
    }
  }

  return false
}

export const getGroupForUser = async (id: UserWithOutPassword['id']) => {
  for (const key of await dataStore.all('group')) {
    const res = await getWrapper(key)

    if (res != null && res.users.some(user => user.id === id)) {
      return res
    }
  }

  return null
}

export const newGroup = async ({
  name = '',
  startSum = 1000,
  smallBlind = 2,
  bigBlind = 5,
  userID,
}: WithOptional<Pick<Group, 'name' | 'startSum'>, 'name' | 'startSum'> & {
  userID: UserWithOutPassword['id']
  smallBlind?: Group['blind']['small']
  bigBlind?: Group['blind']['big']
}): Promise<MaybeNull<Group>> => {
  if (await checkIfAlreadyInAGroup(userID)) {
    return null
  }

  const group = await dataStore.create<Group>('group', id => {
    return {
      id,
      name,
      startSum,
      blind: {
        small: smallBlind,
        big: bigBlind,
      },
      owner: userID,
      users: [
        {
          id: userID,
          sum: startSum,
        },
      ],
    }
  })

  return group
}

const updateWrapper = async (
  id: User['id'],
  breakIfTruthy: (result: Group) => boolean,
  modify: (result: Group) => Promise<Group>
): Promise<MaybeNull<Group>> => {
  const res = await getWrapper(id)

  if (res == null || breakIfTruthy(res)) {
    return null
  }

  const modified = await modify(clone(res))

  await dataStore.update(id, modified, 'group')

  return await modified
}

export const joinGroup = async ({
  id,
  userID,
}: Pick<User, 'id'> & { userID: UserWithOutPassword['id'] }): Promise<
  MaybeNull<Group>
> => {
  if (await checkIfAlreadyInAGroup(userID)) {
    return null
  }

  console.log(id, 'user joined', userID)

  return await updateWrapper(
    id,
    res => res.action != null || res.users.some(user => user.id === userID),
    async res => {
      res.users.push({
        id: userID,
        sum: res.startSum,
      })

      return res
    }
  )
}

export const leaveGroup = async ({
  id,
  userID,
}: Pick<User, 'id'> & { userID: UserWithOutPassword['id'] }): Promise<
  MaybeNull<Group>
> => {
  return await updateWrapper(
    id,
    res => res.action != null || res.users.every(user => user.id !== userID),
    async res => {
      const currentIndex = res.users.findIndex(item => item.id == userID)

      if (currentIndex === -1) {
        return res
      }

      res.users.splice(currentIndex, 1)

      res.owner = res.users.length >= 1 ? res.users[0].id : userID

      if (res.users.length === 0) {
        await dataStore.del({ id, type: 'group' })
      }

      return res
    }
  )
}

export const updateOrder = async ({
  id,
  order,
  userID,
}: Pick<Group, 'id'> & {
  order: Order
  userID: UserWithOutPassword['id']
}) => {
  return await updateWrapper(
    id,
    res => res.action != null || res.owner !== userID,
    async res => {
      const max = res.users.length - 1
      for (const [key, id] of Object.entries(order)) {
        const newIndex = parseInt(key, 10)
        const currentIndex = res.users.findIndex(item => item.id == id)

        if (newIndex < 0 || newIndex > max || currentIndex === newIndex) {
          continue
        }

        const pop = res.users.splice(currentIndex, 1).pop()

        if (pop == null) {
          continue
        }

        res.users.splice(newIndex, 0, pop)
      }

      return res
    }
  )
}

export const updateGroup = async ({
  id,
  name,
  startSum,
  smallBlind,
  bigBlind,
  owner,
  userID,
}: Pick<Group, 'id'> & {
  name: MaybeUndefined<Group['name']>
  startSum: MaybeUndefined<Group['startSum']>
  smallBlind: MaybeUndefined<Group['blind']['small']>
  bigBlind: MaybeUndefined<Group['blind']['big']>
  owner: MaybeUndefined<Group['owner']>
  userID: UserWithOutPassword['id']
}): Promise<MaybeNull<Group>> => {
  const update = {
    bigBlind: bigBlind != null,
    smallBlind: smallBlind != null,
    name: name != null,
    owner: owner != null,
    startSum: startSum != null,
  }

  return await updateWrapper(
    id,
    res => {
      // I am not proud of this part
      if (res.action != null || res.owner !== userID) {
        return true
      }
      if (bigBlind && res.blind.big === bigBlind) {
        update.bigBlind = false
      }
      if (smallBlind && res.blind.small === smallBlind) {
        update.smallBlind = false
      }
      if (name && res.name === name) {
        update.name = false
      }
      if (owner && res.users.find(({ id }) => id === owner) == null) {
        update.owner = false
      }
      if (startSum && res.startSum === startSum) {
        update.startSum = false
      }

      if (update.bigBlind && bigBlind) {
        const small = smallBlind ? smallBlind : res.blind.small
        const start = startSum ? startSum : res.startSum
        update.bigBlind = bigBlind > small && bigBlind < start
      }

      if (update.smallBlind && smallBlind) {
        const big = bigBlind ? bigBlind : res.blind.big
        const start = startSum ? startSum : res.startSum
        update.smallBlind = smallBlind < big && smallBlind < start
      }

      if (update.startSum && startSum) {
        const big = bigBlind ? bigBlind : res.blind.big
        update.startSum = startSum > big
      }

      if (
        update.bigBlind &&
        update.smallBlind &&
        update.startSum &&
        bigBlind &&
        smallBlind &&
        startSum
      ) {
        if (
          bigBlind <= smallBlind ||
          (bigBlind >= startSum || smallBlind >= startSum) ||
          startSum <= bigBlind
        ) {
          update.bigBlind = false
          update.smallBlind = false
          update.startSum = false
        }
      } else if (
        update.bigBlind &&
        update.smallBlind &&
        bigBlind &&
        smallBlind
      ) {
        if (
          bigBlind <= smallBlind ||
          (bigBlind >= res.startSum || smallBlind >= res.startSum)
        ) {
          update.bigBlind = false
          update.smallBlind = false
        }
      } else if (update.bigBlind && update.startSum && bigBlind && startSum) {
        if (
          bigBlind >= startSum ||
          bigBlind <= res.blind.small ||
          startSum <= res.blind.small
        ) {
          update.bigBlind = false
          update.startSum = false
        }
      } else if (
        update.smallBlind &&
        update.startSum &&
        smallBlind &&
        startSum
      ) {
        if (
          smallBlind <= startSum ||
          smallBlind >= res.blind.big ||
          startSum <= res.blind.big
        ) {
          update.smallBlind = false
          update.startSum = false
        }
      }

      return Object.values(update).every(value => value === false)
    },
    async res => {
      if (update.bigBlind && bigBlind) {
        res.blind.big = bigBlind
      }
      if (update.smallBlind && smallBlind) {
        res.blind.small = smallBlind
      }
      if (update.name && name) {
        res.name = name
      }
      if (update.owner && owner) {
        res.owner = owner
      }
      if (update.startSum && startSum) {
        res.startSum = startSum

        res.users.forEach(user => {
          user.sum = startSum
        })
      }
      return res
    }
  )
}

export const startGame = async ({
  id,
  userID,
}: Pick<Group, 'id'> & {
  userID: UserWithOutPassword['id']
}): Promise<MaybeNull<Group>> => {
  return await updateWrapper(
    id,
    res => res.owner !== userID || res.users.length < 2,
    async res => {
      const deck = newDeck()
      const [first, smallBlind] = res.users
      const [, , bigBlind = first] = res.users

      res.users.forEach(user => {
        user.sum = res.startSum
      })

      const turn: ActionRunning['turn'] = {
        [smallBlind.id]: { bet: res.blind.small, status: 'none' },
        [bigBlind.id]: { bet: res.blind.big, status: 'none' },
      }

      res.users
        .filter(({ id }) => id !== bigBlind.id && id !== smallBlind.id)
        .forEach(user => {
          turn[user.id] = {
            bet: 0,
            status: 'none',
          }
        })

      smallBlind.sum -= res.blind.small
      bigBlind.sum -= res.blind.big

      const action = await dataStore.create<ActionRunning>('action', id => ({
        id,
        groupID: res.id,
        queued: {},
        button: smallBlind.id,
        big: bigBlind.id,
        small: smallBlind.id,
        turn,
        deck,
        communityCards: [],
        pot: res.blind.small + res.blind.big,
        round: 0,
      }))

      if (action == null) {
        throw new Error('atomic update failed')
      }

      res.action = action.id

      console.log(action.id, 'new action', res.users.map(user => user.id))

      return res
    }
  )
}
