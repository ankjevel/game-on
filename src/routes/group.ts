import { Route } from 'Route'
import { isNumber, hasProp, nullOrEmpty, parse } from '../utils'
import * as groupService from '../services/group'
import { UserWithOutPassword, Group } from 'dataStore'

const input = {
  name(input: any): MaybeUndefined<Group['name']> {
    return hasProp<Group>(input, 'name') &&
      typeof input.name === 'string' &&
      nullOrEmpty(input.name) === false
      ? input.name.substr(0, 255)
      : undefined
  },

  startSum(input: any): MaybeUndefined<Group['startSum']> {
    return hasProp<Group>(input, 'startSum') &&
      isNumber(input.startSum) &&
      isNumber(parseInt((input as any).startSum))
      ? Math.max(
          0,
          Math.min(parseInt((input as any).startSum), Number.MAX_SAFE_INTEGER)
        )
      : undefined
  },

  owner(
    input: any,
    userID: UserWithOutPassword['id']
  ): MaybeUndefined<Group['owner']> {
    return hasProp<Group>(input, 'owner') &&
      typeof input.owner === 'string' &&
      nullOrEmpty(input.owner) === false &&
      userID !== input.owner
      ? input.owner
      : undefined
  },

  order(order: any): MaybeNull<{ [order: string]: string }> {
    if (
      order == null ||
      Object.keys(order).every(key => isNumber(key)) === false ||
      Object.values(order).every(
        value => !nullOrEmpty(value) && parse(value as string) == null
      ) === false
    ) {
      return null
    }

    return order
  },

  param(param: string) {
    return nullOrEmpty(param) || param.length > 255
  },
}

export const register: Route = (app, auth) => {
  app.post('/group', auth, async ({ body, user }, res) => {
    if (user == null) {
      return res.sendStatus(400)
    }

    res.send(
      await groupService.newGroup({
        name: input.name(body),
        startSum: input.startSum(body),
        userID: user.id,
      })
    )
  })

  app.put('/group/:id/join', auth, async ({ params: { id }, user }, res) => {
    if (user == null || input.param(id)) {
      return res.sendStatus(400)
    }

    const result = await groupService.joinGroup({ id, userID: user.id })

    if (result == null) {
      return res.sendStatus(400)
    }

    res.send(result)
  })

  app.delete(
    '/group/:id/leave',
    auth,
    async ({ params: { id }, user }, res) => {
      if (user == null || input.param(id)) {
        return res.sendStatus(400)
      }

      const result = await groupService.leaveGroup({ id, userID: user.id })
      if (result == null) {
        return res.sendStatus(400)
      }

      res.send({ status: `left "${result.name}"` })
    }
  )

  app.delete('/group/:id', auth, async ({ params: { id }, user }, res) => {
    if (user == null || input.param(id)) {
      return res.sendStatus(400)
    }

    const result = await groupService.deleteGroup({ id, userID: user.id })
    if (result == null) {
      return res.sendStatus(400)
    }

    res.send(result)
  })

  app.patch('/group/:id', auth, async ({ params: { id }, body, user }, res) => {
    if (user == null || input.param(id)) {
      return res.sendStatus(400)
    }

    const result = await groupService.updateGroup({
      id,
      owner: input.owner(body, user.id),
      name: input.name(body),
      startSum: input.startSum(body),
      userID: user.id,
    })

    if (result == null) {
      return res.sendStatus(400)
    }

    res.send(result)
  })

  app.patch(
    '/group/:id/order',
    auth,
    async ({ params: { id }, body, user }, res) => {
      const order = input.order(body)

      if (user == null || input.param(id) || order == null) {
        return res.sendStatus(400)
      }

      const result = await groupService.updateOrder({
        id,
        order,
        userID: user.id,
      })

      if (result == null) {
        return res.sendStatus(400)
      }

      res.send(result)
    }
  )

  app.put('/group/:id/start', auth, async ({ params: { id }, user }, res) => {
    if (user == null || input.param(id)) {
      return res.sendStatus(400)
    }

    const result = await groupService.startGame({ id, userID: user.id })

    if (result == null) {
      return res.sendStatus(400)
    }

    res.send(result)
  })
}
