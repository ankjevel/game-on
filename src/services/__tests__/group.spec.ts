import * as dataStore from '../dataStore'

jest.mock('../dataStore')

import * as group from '../group'

let ds: {
  [key: string]: jest.Mock
}

beforeEach(() => {
  ds = dataStore as any
  console.log = jest.fn()
})

describe('#startGame', () => {
  it('works', async () => {
    const id = 'foo'
    const userID = 'bar'
    const res = {
      id,
      owner: userID,
      turn: {},
      startSum: 200,
      blind: { small: 2, big: 4 },
      users: [{ id: userID }, { id: 'baz' }],
    }

    ds.get.mockResolvedValue(res)
    ds.update.mockResolvedValue(res)

    let action
    ds.create.mockImplementation((type, callback) => {
      action = callback(`${type}:6e002e4c-c194-42ab-aa15-f0aa6a8215e3`)
      return action
    })

    const game = await group.startGame({ id, userID })

    expect(game).toMatchSnapshot()
    expect(action).toMatchSnapshot()
  })
})
