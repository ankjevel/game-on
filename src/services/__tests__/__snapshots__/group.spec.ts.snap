// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#startGame works 1`] = `
Object {
  "action": "action:6e002e4c-c194-42ab-aa15-f0aa6a8215e3",
  "blind": Object {
    "big": 4,
    "small": 2,
  },
  "id": "foo",
  "owner": "bar",
  "startSum": 200,
  "turn": Object {},
  "users": Array [
    Object {
      "id": "bar",
      "sum": 196,
    },
    Object {
      "id": "baz",
      "sum": 198,
    },
  ],
}
`;

exports[`#startGame works 2`] = `
Object {
  "big": "bar",
  "button": "baz",
  "communityCards": Array [],
  "deck": Array [
    "1F0A1",
    "1F0A2",
    "1F0A3",
    "1F0A4",
    "1F0A5",
    "1F0A6",
    "1F0A7",
    "1F0A8",
    "1F0A9",
    "1F0A10",
    "1F0A11",
    "1F0A12",
    "1F0A13",
    "1F0B1",
    "1F0B2",
    "1F0B3",
    "1F0B4",
    "1F0B5",
    "1F0B6",
    "1F0B7",
    "1F0B8",
    "1F0B9",
    "1F0B10",
    "1F0B11",
    "1F0B12",
    "1F0B13",
    "1F0C1",
    "1F0C2",
    "1F0C3",
    "1F0C4",
    "1F0C5",
    "1F0C6",
    "1F0C7",
    "1F0C8",
    "1F0C9",
    "1F0C10",
    "1F0C11",
    "1F0C12",
    "1F0C13",
    "1F0D1",
    "1F0D2",
    "1F0D3",
    "1F0D4",
    "1F0D5",
    "1F0D6",
    "1F0D7",
    "1F0D8",
    "1F0D9",
    "1F0D10",
    "1F0D11",
    "1F0D12",
    "1F0D13",
  ],
  "groupID": "foo",
  "id": "action:6e002e4c-c194-42ab-aa15-f0aa6a8215e3",
  "pot": 6,
  "queued": Object {},
  "round": 0,
  "small": "baz",
  "turn": Object {
    "bar": Object {
      "bet": 4,
      "status": "none",
    },
    "baz": Object {
      "bet": 2,
      "status": "none",
    },
  },
}
`;
